filebeat/input/filestream/input_test.go

```diff
@@ -0,0 +1,332 @@
+func TestFilestreamCloseRenamed(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "TestFilestreamCloseRenamed")
+	if err != nil {
+		t.Fatalf("cannot create temporary folder for test file: %+v", err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	testlogPath := filepath.Join(tmpDir, "test.log")
+	err = ioutil.WriteFile(testlogPath, []byte("first log line\n"), 0644)
+	if err != nil {
+		t.Fatalf("cannot write log lines to test file: %+v", err)
+	}
+
+	testStore := openTestStatestore(t, tmpDir)
+	inp := getInputFromConfig(
+		t,
+		testStore,
+		map[string]interface{}{
+			"paths":                                []string{testlogPath},
+			"prospector.scanner.check_interval":    "1s",
+			"close.on_state_change.check_interval": "1s",
+			"close.on_state_change.renamed":        "true",
+		},
+	)
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	inputCtx := input.Context{Logger: logp.L(), Cancelation: ctx}
+	pipeline := &mockPipelineConnector{}
+
+	var wg sync.WaitGroup
+	go func() {
+		wg.Add(1)
+		defer wg.Done()
+
+		inp.Run(inputCtx, pipeline)
+	}()
+
+	// first event has made it successfully
+	waitUntilEventCount(1, pipeline)
+	// check registry
+	checkOffsetInRegistry(t, testStore, testlogPath, len([]byte("first log line\n")))
+
+	testlogPathRenamed := filepath.Join(tmpDir, "test.log.rotated")
+	err = os.Rename(testlogPath, testlogPathRenamed)
+	if err != nil {
+		t.Fatalf("failed to rename file '%s': %+v", testlogPath, err)
+	}
+
+	// wait enough time so renaming of file can be detected
+	time.Sleep(2 * time.Second)
+
+	err = ioutil.WriteFile(testlogPath, []byte("new first log line\nnew second log line\n"), 0644)
+	if err != nil {
+		t.Fatalf("cannot write log lines to test file: %+v", err)
+	}
+
+	// new two events are read
+	waitUntilEventCount(2, pipeline)
+
+	cancelInput()
+	wg.Wait()
+
+	checkOffsetInRegistry(t, testStore, testlogPathRenamed, len([]byte("first log line\n")))
+	checkOffsetInRegistry(t, testStore, testlogPath, len([]byte("new first log line\nnew second log line\n")))
+}
+
+// test_close_eof from test_harvester.py
+func TestFilestreamCloseEOF(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "TestFilestreamCloseEOF")
+	if err != nil {
+		t.Fatalf("cannot create temporary folder for test file: %+v", err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	testlogPath := filepath.Join(tmpDir, "test.log")
+	err = ioutil.WriteFile(testlogPath, []byte("first log line\n"), 0644)
+	if err != nil {
+		t.Fatalf("cannot write log lines to test file: %+v", err)
+	}
+
+	testStore := openTestStatestore(t, tmpDir)
+	inp := getInputFromConfig(
+		t,
+		testStore,
+		map[string]interface{}{
+			"paths":                             []string{testlogPath},
+			"prospector.scanner.check_interval": "24h",
+			"close.reader.on_eof":               "true",
+		},
+	)
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	inputCtx := input.Context{Logger: logp.L(), Cancelation: ctx}
+	pipeline := &mockPipelineConnector{}
```

Comment:
these tests have quite a few moving pieces/dependencies. Namely: directory setup, statestore setup, pipeline setup. Maybe we also want to capture log messages. You might want to consider to create a single `type inputsTestEnvironment struct{...}` with a set of methods in order to encapsulate the boilerplate. e.g. instead of `ioutil.WriteFile` we could write `env.disk.WriteFile(path, contents)`. Instead of `checkOffsetInRegistry` we could write `env.registry.CheckOffset(path, len)`.